Update toolchain: GCC 8.4 (C ++ 17 standard), crosstool-ng 1.24, add MIPS64, PPC32, PPC64 and others things

Bensuperpc opened the pull request:
Hello,
I updated the GCC compiler to version 8.3.0 for C ++ 17 support, on ARMv7, ppc64le and mips, and GCC 7.5.0 for ARMv5.
There are the other images that need to be updated, I will do it later if I have the time.

Bensuperpc commented:
I don't know why the manylinux isn't working, yet I haven't touched their config files or haven't seen something ?

Bensuperpc commented:
I don't know why the manylinux isn't working, yet I haven't touched their config files or haven't seen something ?

JonasVautherin commented:
Seems unrelated :thinking::

```
Installing collected packages: pip
Successfully installed pip-21.1.1
Traceback (most recent call last):
  File "/opt/python/cp38-cp38/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/python/cp38-cp38/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/pip/__main__.py", line 23, in <module>
    from pip._internal.cli.main import main as _main  # isort:skip # noqa
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/pip/_internal/cli/main.py", line 10, in <module>
    from pip._internal.cli.autocompletion import autocomplete
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/pip/_internal/cli/autocompletion.py", line 9, in <module>
    from pip._internal.cli.main_parser import create_main_parser
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/pip/_internal/cli/main_parser.py", line 7, in <module>
    from pip._internal.cli import cmdoptions
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/pip/_internal/cli/cmdoptions.py", line 25, in <module>
    from pip._internal.cli.progress_bars import BAR_TYPES
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/pip/_internal/cli/progress_bars.py", line 12, in <module>
    from pip._internal.utils.logging import get_indentation
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/pip/_internal/utils/logging.py", line 18, in <module>
    from pip._internal.utils.misc import ensure_dir
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/pip/_internal/utils/misc.py", line 34, in <module>
    from pip._internal.locations import get_major_minor_version, site_packages, user_site
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/pip/_internal/locations/__init__.py", line 9, in <module>
    from . import _distutils, _sysconfig
  File "/opt/python/cp38-cp38/lib/python3.8/site-packages/pip/_internal/locations/_sysconfig.py", line 8, in <module>
    from pip._internal.exceptions import InvalidSchemeCombination, UserInstallationInvalid
ImportError: cannot import name 'InvalidSchemeCombination' from 'pip._internal.exceptions' (/opt/python/cp38-cp38/lib/python3.8/site-packages/pip/_internal/exceptions.py)
The command '/bin/sh -c X86_FLAG=$([ "$DEFAULT_DOCKCROSS_IMAGE" = "dockcross/manylinux1-x86" -o "$DEFAULT_DOCKCROSS_IMAGE" = "dockcross/manylinux2010-x86" -o "$DEFAULT_DOCKCROSS_IMAGE" = "dockcross/manylinux2014-x86" ] && echo "-32" || echo "") &&   /buildscripts/build-and-install-openssl.sh $X86_FLAG &&   /buildscripts/build-and-install-openssh.sh &&   /buildscripts/build-and-install-curl.sh &&   /buildscripts/build-and-install-git.sh &&   /buildscripts/install-cmake-binary.sh $X86_FLAG &&   /buildscripts/install-liquidprompt-binary.sh &&   PYTHON=$([ -e /opt/python/cp38-cp38/bin/python ] && echo "/opt/python/cp38-cp38/bin/python" || echo "python") &&   /buildscripts/install-python-packages.sh -python ${PYTHON} &&   /buildscripts/build-and-install-ninja.sh -python ${PYTHON} &&   rm -rf /buildscripts' returned a non-zero code: 1
make: *** [Makefile:231: manylinux1-x64] Error 1

Error: Process completed with exit code 2.
```

JonasVautherin commented:
I see that you added a new workflow. Would you mind elaborating on it? Why is it required? Why do we need it to run on each push on each branch instead of the current `main.yml` workflow? :thinking:

Bensuperpc commented:
This is a temporary workflow, i don't know if it is possible to have it only on my git repository.
the advantage of doing this on each push, it allows me to see if a modification works on all the images and it is faster than on my PC.
It is a habit that I have with my computer school, to check if everything works with each push thanks to CI.

Bensuperpc commented:
Currently I am updating the toolchain rather than just GCC, I have updated the title.

Bensuperpc commented:
I updated a lot of things (On Linux image only):

- GCC to 8.4 (Add support C++14 and C++17) #491 
- CMake to 3.20.2
- glibc to 2.29
- binutils to 2.32
- Crosstool-ng to 2.24 (And update conf)
- expat to 2.3
- mpfr to 4.1
- Enable quadmathlib
- Update to stable-20210408-slim (From stretch-20190326-slim)
- Add Docker script to add Boost and OpenCV libs (Only tested on Mips64 for now)
- Remove old images manylinux1 and manylinux2010 (EOL, but i noticed this #458 after i removed image, maybe put back manylinux2010 and manylinux1 ?)
- Add Mips64  (#476), m68K, ppc64, ppc32... 
- Add ARMv6-rpi image with newer version of GCC 8.4.0 (Official toolchain still on GCC 4.9)
- Move ARM64 image to armv8 (ARMv9 is coming soon)
I hope I did not forget anything, if you have any suggestions, don't hesitate ^^
There is still manylinux 2014-aarch64 which still has some problems (because of too old GNU make) :/

JonasVautherin commented:
Thanks for the contributions!

However, it looks to me like those should be split in smaller PRs. For instance "Update CI workflow" could be a PR on its own. The version updates on binaries (binutils, glibc, ...) could be another one, the GCC update would be another one, etc.

The problem now is that there are some things that I could review and merge, but some for which I need @thewtex or @jcfr to have a look. And because this is all in one huge PR, it makes it very difficult to review.

Does that make sense?

thewtex commented:
@Bensuperpc Wow! Amazing work! :clap: :rocket: :star2: 

I know it is more work, but as @JonasVautherin said, breaking these into multiple PR's will help review and keep the builds green.

Bensuperpc commented:
I will divide this into multiple pull requests ^^
